master/internal/resourcemanagers/resource_pool.go

```diff
@@ -163,31 +177,45 @@ func (rp *ResourcePool) allocateResources(ctx *actor.Context, req *sproto.Alloca
 		case error:
 			// Rollback previous allocations.
 			ctx.Log().WithError(resp).Warnf("failed to allocate request %s", req.AllocationID)
-			for _, resource := range resources {
-				resource := resource.(*containerResources)
-				ctx.Tell(resource.agent.Handler,
-					agent.DeallocateContainer{ContainerID: resource.containerID})
-			}
-
+			rollback = true
 			return false
 		default:
 			panic(fmt.Sprintf("bad AllocateFreeDevices response: %s", resp))
 		}
 	}
 
+	// Persist allocation_resources and container_resources.
+	for _, cr := range resources {
+		rs := task.NewResourcesState(cr, -1)
+		if err := rs.Persist(); err != nil {
```

Comment:
seems better to me to do all the db persists then the agent ones. so you can't have a container on an agent that the master doesn't know about if it crashes.

Reply:
I think I misread here.